### PR TITLE
Adding grid functionalities for plotting

### PR DESCRIPTION
**Pull request by jemorlanes**
Motivated by VisiumHD, also included a vignette with explanations

### CODE REVIEW

File: R/visualization_spatial.R [hunk 1]
```diff
@@ -63,6 +68,9 @@ NULL
 #' visible in small plots.
 #' @param scalebar_position A numeric vector of length 2 specifying the position of the scale bar
 #' relative to the plot area. Default is to place it in the top right corner.
+#' @param tech Relevant for tile and raster shapes. A string indicating which technology the data comes from. 
+#' Defaults to \code{NULL} and will detect the technology automatically from the data. 
+#' Possible values are \code{c("vis", "vishd")}
```
Comment (ludvigla) [inline R/visualization_spatial.R:73]: Might be a good idea to spell these out for clarity. E.g. "visium" and "visiumhd"

File: R/visualization_spatial.R [hunk 2]
```diff
@@ -125,16 +136,17 @@ MapFeatures.default <- function (
   }
 
   # get features
-  features <- object |>
-    select(-barcode, -all_of(coords_columns), -sampleID, -all_of(label_by)) |>
-    colnames()
-
+  features <- setdiff(colnames(object), c("barcode", coords_columns, "sampleID", label_by, "pxl_col_in_fullres", "pxl_row_in_fullres"))
+  # features <- object |>
```
Comment (ludvigla) [inline R/visualization_spatial.R:140]: Remove commented code

File: R/visualization_spatial.R [hunk 3]
```diff
@@ -612,6 +727,8 @@ MapLabels.default <- function (
   pt_size = 1,
   pt_alpha = 1,
   pt_stroke = 0,
+  shape = "point",
+  spot_side,
```
Comment (ludvigla) [inline R/visualization_spatial.R:731]: Should this have a default value?

File: R/visualization_spatial.R [hunk 4]
```diff
@@ -635,9 +753,10 @@ MapLabels.default <- function (
   .prep_data_for_plotting(object = object, label_by = label_by, coords_columns = coords_columns)
 
   # get label column
-  label <- object |>
-    select(-barcode, -all_of(coords_columns), -sampleID, -all_of(label_by)) |>
-    colnames()
+  label <- setdiff(colnames(object), c("barcode", coords_columns, "sampleID", label_by, "pxl_col_in_fullres", "pxl_row_in_fullres"))
+  # label <- object |>
```
Comment (ludvigla) [inline R/visualization_spatial.R:757]: Remove comments

File: R/visualization_spatial.R [hunk 5]
```diff
@@ -1292,6 +1501,518 @@ MapLabels.Seurat <- function (
...
+  } else {
+    abort(glue("Available plotting shapes are {col_br_green('raster')} or {col_br_green('tile')}"))
+  }
+
+  p <- p +
+    {
+      if((shape == "raster" | shape == "tile") & "xy" == paste(coords_columns, collapse = "")){
+        # Set plot dimensions (adjust for array coordinates used in raster)
+        x_lim <- max(gg$x) + 1
+        # if(max(gg$y) <= 77) { # 6.5x6.5 mm arrays, max y coord is 77
```
Comment (ludvigla) [inline R/visualization_spatial.R:1685]: Remove comments

File: R/visualization_spatial.R [hunk 6]
```diff
@@ -1367,18 +2088,28 @@ MapLabels.Seurat <- function (
...
   } else {
     if (!multi_color) {
+      remain <- setdiff(colnames(object), c("barcode", coords_columns, "sampleID", label_by, "pxl_col_in_fullres", "pxl_row_in_fullres"))
       checks <- object |>
-        select(-barcode, -all_of(coords_columns), -sampleID, -all_of(label_by)) |>
+        select(all_of(remain)) |>
         sapply(function(x) {
           is.character(x) | is.factor(x)
         })
+      # checks <- object |>
```
Comment (ludvigla) [inline R/visualization_spatial.R:2108]: Remove comments

File: tests/testthat/test-MapFeatures.R [hunk 7]
```diff
@@ -13,69 +13,218 @@ se_merged <- MergeSTData(se_mbrain, se_mcolon) |>
...
   expect_true(all(c("Clu", "Slc6a3", "Vip") == names(p[[1]])))
   
-  # Test that the expected number of plots are produced
+  ## Test that the expected number of plots are produced
   p <- MapFeatures(se_merged, features = c("Clu", "Slc6a3", "Vip"))
   expect_equal(p$patches$layout$ncol, 3)
+  
+  # Tile shape
+  ## Test image_use parameter
+  expect_s3_class(MapFeatures(se_merged, features = c("Clu", "Slc6a3", "Vip"), image_use = "raw", shape = "tile"), "patchwork")
```
Comment (ludvigla) [inline tests/testthat/test-MapFeatures.R:83]: Preferably, the tests should make sure that the content of the plots is as expected. It is possible to extract parts of the patchwork object to verify that a parameter setting works as expected. Just checking the class is not going to be very helpful on its own.

**Review by ludvigla — APPROVED** [review]

**Comment by ludvigla** [comment]
Look really good! If possible, it might be worthwhile thinking about how some of the code can be refactored to trim it down a bit. Some parts are used multiple times and could possible be moved to utility functions instead. I'm probably responsible for most of those examples, but feel free to e.g. replace complex if/else statements with `switch`.

Also, make sure to update the changelog before merging!

File: vignettes/.DS_Store
Comment (ludvigla) [inline vignettes/.DS_Store:1]: Remove this file!

File: tests/testthat/Rplots.pdf
Comment (ludvigla) [inline tests/testthat/Rplots.pdf:1]: This should not be included!

File: R/visualization_spatial.R (hunk 3 above)
Comment (jemorlanes) [reply R/visualization_spatial.R:731]: The value for `spot_side` is assigned when calling `MapLabels.default`, as we retrieve `spot_side` with `MapLabels.Seurat`. In `MapLabels.Seurat`, the default value is NULL. Should I do the same for the default value in `MapLabels.default`?

Comment (ludvigla) [reply R/visualization_spatial.R:728]: Not so sure, but if you think it makes more sense to not have a default I would move up the argument to the top. It's good practice to put required arguments on the top and then parameters with defaults at the bottom.

File: R/visualization_spatial.R (hunk 1 above)
Comment (jemorlanes) [reply R/visualization_spatial.R:73]: Done. I have checked the code properly and there was no real need for a `tech` parameter. Thus, it has been removed and code related to it aswell

File: R/visualization_spatial.R (hunk 3 above)
Comment (jemorlanes) [reply R/visualization_spatial.R:728]: Argument has now been moved to the top of the function for the `.default` functions. In `.Seurat`, they stay with NULL.

**Pull request merged**